subprojects/core/src/main/java/org/gradle/internal/configuration/BuildOperationDomainObjectConfigurator.java

```diff
@@ -0,0 +1,81 @@
+import org.apache.commons.lang.StringUtils;
+import org.gradle.api.Action;
+import org.gradle.api.Nullable;
+import org.gradle.internal.operations.BuildOperationContext;
+import org.gradle.internal.progress.BuildOperationDetails;
+import org.gradle.internal.progress.BuildOperationExecutor;
+import org.gradle.internal.progress.HasBuildOperationContext;
+
+/**
+ * A decorating {@link DomainObjectConfigurator} implementation that delegates to a given
+ * decorated implementation, but wraps the configure() execution in a
+ * {@link org.gradle.internal.operations.BuildOperation}. The wrapping is only performed
+ * if the domain object under configuration is implementing {@link HasBuildOperationContext}.
+ */
+public class BuildOperationDomainObjectConfigurator implements DomainObjectConfigurator {
+
+    private DomainObjectConfigurator decorated;
+    private BuildOperationExecutor buildOperationExecutor;
+
+    public BuildOperationDomainObjectConfigurator(DomainObjectConfigurator decorated, BuildOperationExecutor buildOperationExecutor) {
+        this.decorated = decorated;
+        this.buildOperationExecutor = buildOperationExecutor;
+    }
+
+    @Override
+    public <T> T configure(final T object, final Closure configureClosure, @Nullable final String callerContextInformation) {
+        if (object instanceof HasBuildOperationContext) {
+            buildOperationExecutor.run(toDisplayName((HasBuildOperationContext) object, callerContextInformation), new Action<BuildOperationContext>() {
+                @Override
+                public void execute(BuildOperationContext buildOperationContext) {
+                    decorated.configure(object, configureClosure, callerContextInformation);
+                }
+            });
+        } else {
+            decorated.configure(object, configureClosure, callerContextInformation);
+        }
+        return object;
+    }
+
+    @Override
+    public <T> T configure(final T object, final Action<? super T> configureAction, @Nullable final String callerContextInformation) {
+        if (object instanceof HasBuildOperationContext) {
+            buildOperationExecutor.run(toDisplayName((HasBuildOperationContext) object, callerContextInformation), new Action<BuildOperationContext>() {
+                @Override
+                public void execute(BuildOperationContext buildOperationContext) {
+                    decorated.configure(object, configureAction, callerContextInformation);
+                }
+            });
+        } else {
+            decorated.configure(object, configureAction, callerContextInformation);
+        }
+        return object;
+    }
+
+    private BuildOperationDetails toDisplayName(HasBuildOperationContext objectWithIdentityPath, String callerContextInformation) {
+        String nameWithContext = objectWithIdentityPath.getBuildOperationDisplayName();
+        if (callerContextInformation != null) {
+            nameWithContext += " (" + callerContextInformation + ")";
```

Comment:
We don't want to concatenate the context with the object, we just want nested build operations. I.e. there should be a build operation for the whole `allProjects {}` call and then a build operation for `Configuring project X` below it.

```
Configuring root
  Executing `allprojects {}` action
    Configuring sub1
      applying plugin foo
    Configuring sub2
      applying plugin foo
...
```